demo/core_13_$.utils.code.js

```diff
@@ -137,64 +137,111 @@ $.utils.code.toSourceSafe = function(value) {
-      // This is an object literal: {a: 1}
-      return '(' + src + '\n)';
+    return src;
+  }
+  // Try parsing src as an expression.
+  // This may throw.
+  ast = $.utils.acorn.parseExpressionAt(src, 0);
+  var remainder = src.substring(ast.end).trim();
+  if (remainder !== '') {
+    // Trim off any trailing comments.
+    var remainderAst = null;
+    try {
+      remainderAst = $.utils.acorn.parse(remainder);
+    } catch (e) {
+      // remainderAst remains null.
+    }
+    if (!remainderAst) {
+      throw SyntaxError('Syntax error beyond expression');
+    }
+    if (remainderAst.type !== 'Program') {
+      throw SyntaxError('Unexpected code beyond expression');  // Module?
+    }
+    // Trim off any unnecessary trailing semicolons.
+    while (remainderAst.body[0] &&
+           remainderAst.body[0].type === 'EmptyStatement') {
+      remainderAst.body.shift();
+    }
+    if (remainderAst.body.length !== 0) {
+      throw SyntaxError('Only one expression expected');
     }
   }
+  src = src.substring(0, ast.end);
+  if (ast.type === 'ObjectExpression') {
+    // {a: 1} needs to be wrapped in parens to avoid being a syntax error.
+    src = '(' + src + ')';
+  }
   return src;
 };
 
 $.utils.code.rewriteForEval.unittest = function() {
   var cases = {
-    '1 + 2': '1 + 2',
-    '1 + 2;\n3 + 4;': '1 + 2;\n3 + 4;',
-    '{}': '({}\n)',
-    '{};\n3 + 4;': '{};\n3 + 4;',
-    '{"a": 1}': '({"a": 1}\n)',
-    '{"a": 1};': '{"a": 1};',
-    '{a: 1}': '({a: 1}\n)',
-    '{a: 1};': '{a: 1};',
-    '{a: 1}  // Comment': '({a: 1}  // Comment\n)',
-    '{a: 1, b: 2}': '({a: 1, b: 2}\n)',
-    '{a: 1; b: 2}': '{a: 1; b: 2}',
-    '{} + []': '{} + []'
+    // Input: [Expression, Statement(s)]
+    '1 + 2': ['1 + 2', '1 + 2'],
+    '2 + 3  // Comment': ['2 + 3', '2 + 3  // Comment'],
+    '3 + 4;': ['3 + 4', '3 + 4;'],
+    '4 + 5; 6 + 7': [SyntaxError, '4 + 5; 6 + 7'],
+    '{}': ['({})', '({})'],
+    '{}  // Comment': ['({})', '({})'],
+    '{};': ['({})', '{};'],
+    '{}; {}': [SyntaxError, '{}; {}'],
+    '{"a": 1}': ['({"a": 1})', '({"a": 1})'],
+    '{"a": 2}  // Comment': ['({"a": 2})', '({"a": 2})'],
+    '{"a": 3};': ['({"a": 3})', '({"a": 3})'],
+    '{"a": 4}; {"a": 4}': [SyntaxError, SyntaxError],
+    '{b: 1}': ['({b: 1})', '({b: 1}\n)'],
+    '{b: 2}  // Comment': ['({b: 2})', '({b: 2}  // Comment\n)'],
+    '{b: 3};': ['({b: 3})', '{b: 3};'],
+    '{b: 4}; {b: 4}': [SyntaxError, '{b: 4}; {b: 4}'],
+    '{} + []': ['{} + []', '{} + []']
   };
+  var actual;
   for (var key in cases) {
     if (!cases.hasOwnProperty(key)) continue;
-    var actual = $.utils.code.rewriteForEval(key);
-    if (actual !== cases[key]) {
-      throw Error('Expected: ' + cases[key] + ' Actual: ' + actual);
+    for (var i = 0; i < 1; i++) {
+      // Alternate between expression and statement evals.
+      try {
+        actual = $.utils.code.rewriteForEval(key, i === 0);
+      } catch (e) {
+        actual = SyntaxError;
```

Comment:
So, a recent-ish episode of TotT strongly discourages table-driven tests where the test loop itself (rather than the function under test) does something different depending on the test input—and in particular, that for table-driven tests advocated for separating out the valid and invalid tests into separate tests.  The valid cases will have a table including input and expected output; the invalid test table can have just the invalid inputs.

I can see that in this case it is desirable to avoid repeating the test data, but the resulting deeply nested driver code is kind of ugly and not that easy to follow.

Reply:
Agreed about the second loop.  I've unrolled the loop, which I think makes it easier to follow, though does introduce repeated code.  Amusingly, this was my original code, but I refactored it into a loop because I assumed I'd get yelled at in the code review.

Reply:
Yeah: the longer version is definitely more readable but the whole try/catch business is now extra-obnoxious.  I don't know if there's a winning plan here.